File: backend/internal/models/subscription.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// SubscriptionStatus represents the status of a subscription
type SubscriptionStatus string

const (
	SubscriptionStatusTrial    SubscriptionStatus = "trial"
	SubscriptionStatusActive   SubscriptionStatus = "active"
	SubscriptionStatusPastDue  SubscriptionStatus = "past_due"
	SubscriptionStatusCanceled SubscriptionStatus = "canceled"
	SubscriptionStatusExpired  SubscriptionStatus = "expired"
)

// SubscriptionPlan represents available subscription plans
type SubscriptionPlan string

const (
	PlanFree    SubscriptionPlan = "free"
	PlanPremium SubscriptionPlan = "premium"
)

// Subscription represents a user's subscription
type Subscription struct {
	ID                uuid.UUID          `json:"id" db:"id"`
	UserID            uuid.UUID          `json:"user_id" db:"user_id"`
	Plan              SubscriptionPlan   `json:"plan" db:"plan"`
	Status            SubscriptionStatus `json:"status" db:"status"`
	StripeCustomerID  *string            `json:"stripe_customer_id" db:"stripe_customer_id"`
	StripeSubscriptionID *string         `json:"stripe_subscription_id" db:"stripe_subscription_id"`
	StripePriceID     *string            `json:"stripe_price_id" db:"stripe_price_id"`
	TrialStartDate    *time.Time         `json:"trial_start_date" db:"trial_start_date"`
	TrialEndDate      *time.Time         `json:"trial_end_date" db:"trial_end_date"`
	CurrentPeriodStart *time.Time        `json:"current_period_start" db:"current_period_start"`
	CurrentPeriodEnd   *time.Time        `json:"current_period_end" db:"current_period_end"`
	CanceledAt        *time.Time         `json:"canceled_at" db:"canceled_at"`
	CreatedAt         time.Time          `json:"created_at" db:"created_at"`
	UpdatedAt         time.Time          `json:"updated_at" db:"updated_at"`
}

// SubscriptionFeatures represents the features available for each plan
type SubscriptionFeatures struct {
	Plan                SubscriptionPlan `json:"plan"`
	MaxPersonalGoals    *int             `json:"max_personal_goals"`    // nil = unlimited
	GroupGoalsAllowed   bool             `json:"group_goals_allowed"`
	AdvancedAnalytics   bool             `json:"advanced_analytics"`
	PrioritySupport     bool             `json:"priority_support"`
	CustomCategories    bool             `json:"custom_categories"`
	DataExport          bool             `json:"data_export"`
	APIAccess           bool             `json:"api_access"`
}

// PaymentMethod represents a user's payment method
type PaymentMethod struct {
	ID               uuid.UUID `json:"id" db:"id"`
	UserID           uuid.UUID `json:"user_id" db:"user_id"`
	StripePaymentMethodID string `json:"stripe_payment_method_id" db:"stripe_payment_method_id"`
	Type             string    `json:"type" db:"type"` // "card", "bank_account", etc.
	Brand            *string   `json:"brand" db:"brand"`
	Last4            *string   `json:"last4" db:"last4"`
	ExpiryMonth      *int      `json:"expiry_month" db:"expiry_month"`
	ExpiryYear       *int      `json:"expiry_year" db:"expiry_year"`
	IsDefault        bool      `json:"is_default" db:"is_default"`
	CreatedAt        time.Time `json:"created_at" db:"created_at"`
	UpdatedAt        time.Time `json:"updated_at" db:"updated_at"`
}

// Invoice represents a billing invoice
type Invoice struct {
	ID               uuid.UUID `json:"id" db:"id"`
	UserID           uuid.UUID `json:"user_id" db:"user_id"`
	SubscriptionID   uuid.UUID `json:"subscription_id" db:"subscription_id"`
	StripeInvoiceID  string    `json:"stripe_invoice_id" db:"stripe_invoice_id"`
	Amount           float64   `json:"amount" db:"amount"`
	Currency         string    `json:"currency" db:"currency"`
	Status           string    `json:"status" db:"status"`
	PeriodStart      time.Time `json:"period_start" db:"period_start"`
	PeriodEnd        time.Time `json:"period_end" db:"period_end"`
	PaidAt           *time.Time `json:"paid_at" db:"paid_at"`
	DueDate          time.Time `json:"due_date" db:"due_date"`
	InvoiceURL       *string   `json:"invoice_url" db:"invoice_url"`
	CreatedAt        time.Time `json:"created_at" db:"created_at"`
}

// SubscriptionUsage represents usage metrics for billing
type SubscriptionUsage struct {
	ID             uuid.UUID `json:"id" db:"id"`
	UserID         uuid.UUID `json:"user_id" db:"user_id"`
	SubscriptionID uuid.UUID `json:"subscription_id" db:"subscription_id"`
	PersonalGoals  int       `json:"personal_goals" db:"personal_goals"`
	GroupGoals     int       `json:"group_goals" db:"group_goals"`
	GroupsJoined   int       `json:"groups_joined" db:"groups_joined"`
	APIRequests    int       `json:"api_requests" db:"api_requests"`
	StorageUsed    int64     `json:"storage_used" db:"storage_used"` // in bytes
	PeriodStart    time.Time `json:"period_start" db:"period_start"`
	PeriodEnd      time.Time `json:"period_end" db:"period_end"`
	CreatedAt      time.Time `json:"created_at" db:"created_at"`
}

// Request/Response models
type CreateSubscriptionRequest struct {
	Plan              SubscriptionPlan `json:"plan" validate:"required"`
	PaymentMethodID   *string          `json:"payment_method_id,omitempty"`
	PromotionCode     *string          `json:"promotion_code,omitempty"`
}

type UpdateSubscriptionRequest struct {
	Plan *SubscriptionPlan `json:"plan,omitempty"`
}

type CancelSubscriptionRequest struct {
	CancelAtPeriodEnd bool   `json:"cancel_at_period_end"`
	CancellationReason *string `json:"cancellation_reason,omitempty"`
}

type AddPaymentMethodRequest struct {
	PaymentMethodID string `json:"payment_method_id" validate:"required"`
	SetAsDefault    bool   `json:"set_as_default"`
}

type SubscriptionSummary struct {
	Subscription Subscription         `json:"subscription"`
	Features     SubscriptionFeatures `json:"features"`
	Usage        *SubscriptionUsage   `json:"usage,omitempty"`
	NextInvoice  *Invoice             `json:"next_invoice,omitempty"`
}

type BillingPortalRequest struct {
	ReturnURL string `json:"return_url" validate:"required,url"`
}

type BillingPortalResponse struct {
	URL string `json:"url"`
}

// Constructor functions
func NewSubscription(userID uuid.UUID, plan SubscriptionPlan) *Subscription {
	now := time.Now().UTC()
	
	subscription := &Subscription{
		ID:        uuid.New(),
		UserID:    userID,
		Plan:      plan,
		CreatedAt: now,
		UpdatedAt: now,
	}

	// Set up trial for new users
	if plan == PlanPremium {
		subscription.Status = SubscriptionStatusTrial
		subscription.TrialStartDate = &now
		trialEnd := now.AddDate(0, 0, 30) // 30-day trial
		subscription.TrialEndDate = &trialEnd
		subscription.CurrentPeriodStart = &now
		subscription.CurrentPeriodEnd = &trialEnd
	} else {
		subscription.Status = SubscriptionStatusActive
	}

	return subscription
}

func NewPaymentMethod(userID uuid.UUID, stripePaymentMethodID, paymentType string, brand, last4 *string, expiryMonth, expiryYear *int) *PaymentMethod {
	return &PaymentMethod{
		ID:                    uuid.New(),
		UserID:                userID,
		StripePaymentMethodID: stripePaymentMethodID,
		Type:                  paymentType,
		Brand:                 brand,
		Last4:                 last4,
		ExpiryMonth:           expiryMonth,
		ExpiryYear:            expiryYear,
		IsDefault:             false,
		CreatedAt:             time.Now().UTC(),
		UpdatedAt:             time.Now().UTC(),
	}
}

func NewInvoice(userID, subscriptionID uuid.UUID, stripeInvoiceID string, amount float64, currency, status string, periodStart, periodEnd, dueDate time.Time) *Invoice {
	return &Invoice{
		ID:              uuid.New(),
		UserID:          userID,
		SubscriptionID:  subscriptionID,
		StripeInvoiceID: stripeInvoiceID,
		Amount:          amount,
		Currency:        currency,
		Status:          status,
		PeriodStart:     periodStart,
		PeriodEnd:       periodEnd,
		DueDate:         dueDate,
		CreatedAt:       time.Now().UTC(),
	}
}

func NewSubscriptionUsage(userID, subscriptionID uuid.UUID, periodStart, periodEnd time.Time) *SubscriptionUsage {
	return &SubscriptionUsage{
		ID:             uuid.New(),
		UserID:         userID,
		SubscriptionID: subscriptionID,
		PersonalGoals:  0,
		GroupGoals:     0,
		GroupsJoined:   0,
		APIRequests:    0,
		StorageUsed:    0,
		PeriodStart:    periodStart,
		PeriodEnd:      periodEnd,
		CreatedAt:      time.Now().UTC(),
	}
}

// Business logic methods
func (s *Subscription) IsActive() bool {
	return s.Status == SubscriptionStatusActive || s.Status == SubscriptionStatusTrial
}

func (s *Subscription) IsInTrial() bool {
	return s.Status == SubscriptionStatusTrial && s.TrialEndDate != nil && time.Now().UTC().Before(*s.TrialEndDate)
}

func (s *Subscription) IsPremium() bool {
	return s.Plan == PlanPremium && s.IsActive()
}

func (s *Subscription) DaysRemainingInTrial() *int {
	if !s.IsInTrial() {
		return nil
	}
	
	now := time.Now().UTC()
	days := int(s.TrialEndDate.Sub(now).Hours() / 24)
	if days < 0 {
		days = 0
	}
	return &days
}

func (s *Subscription) GetFeatures() SubscriptionFeatures {
	switch s.Plan {
	case PlanPremium:
		return SubscriptionFeatures{
			Plan:                PlanPremium,
			MaxPersonalGoals:    nil, // unlimited
			GroupGoalsAllowed:   true,
			AdvancedAnalytics:   true,
			PrioritySupport:     true,
			CustomCategories:    true,
			DataExport:          true,
			APIAccess:           true,
		}
	default: // PlanFree
		maxGoals := 3
		return SubscriptionFeatures{
			Plan:                PlanFree,
			MaxPersonalGoals:    &maxGoals,
			GroupGoalsAllowed:   false,
			AdvancedAnalytics:   false,
			PrioritySupport:     false,
			CustomCategories:    false,
			DataExport:          false,
			APIAccess:           false,
		}
	}
}

func (s *Subscription) CanCreatePersonalGoal(currentGoalCount int) bool {
	features := s.GetFeatures()
	if features.MaxPersonalGoals == nil {
		return true // unlimited
	}
	return currentGoalCount < *features.MaxPersonalGoals
}

func (s *Subscription) CanJoinGroups() bool {
	return s.GetFeatures().GroupGoalsAllowed
}

func (s *Subscription) HasAdvancedAnalytics() bool {
	return s.GetFeatures().AdvancedAnalytics
}

// Plan pricing constants (in cents)
const (
	PremiumMonthlyPrice = 999  // $9.99/month
	PremiumYearlyPrice  = 9999 // $99.99/year (2 months free)
)

// GetPlanPrice returns the price for a plan in cents
func GetPlanPrice(plan SubscriptionPlan, yearly bool) int {
	switch plan {
	case PlanPremium:
		if yearly {
			return PremiumYearlyPrice
		}
		return PremiumMonthlyPrice
	default:
		return 0
	}
}

File: frontend/src/lib/types/subscription.ts
export type SubscriptionStatus = 'trial' | 'active' | 'past_due' | 'canceled' | 'unpaid';

export type SubscriptionPlan = 'free' | 'premium';

export interface Subscription {
	id: string;
	user_id: string;
	plan: SubscriptionPlan;
	status: SubscriptionStatus;
	stripe_subscription_id?: string;
	stripe_customer_id?: string;
	current_period_start: string;
	current_period_end: string;
	trial_end?: string;
	canceled_at?: string;
	created_at: string;
	updated_at: string;
}

export interface SubscriptionFeatures {
	personal_goals_limit: number;
	can_join_groups: boolean;
	group_limit: number;
	advanced_analytics: boolean;
	export_data: boolean;
	priority_support: boolean;
	custom_categories: boolean;
	api_access: boolean;
}

export interface PaymentMethod {
	id: string;
	user_id: string;
	stripe_payment_method_id: string;
	type: 'card';
	card_brand: string;
	card_last4: string;
	card_exp_month: number;
	card_exp_year: number;
	is_default: boolean;
	created_at: string;
}

export interface Invoice {
	id: string;
	user_id: string;
	subscription_id: string;
	stripe_invoice_id: string;
	amount_paid: number;
	currency: string;
	status: 'draft' | 'open' | 'paid' | 'uncollectible' | 'void';
	invoice_pdf?: string;
	hosted_invoice_url?: string;
	period_start: string;
	period_end: string;
	created_at: string;
}

export interface SubscriptionUsage {
	id: string;
	user_id: string;
	subscription_id: string;
	personal_goals_count: number;
	groups_joined: number;
	data_export_requests: number;
	api_requests: number;
	period_start: string;
	period_end: string;
	created_at: string;
}

// Request/Response types
export interface CreateSubscriptionRequest {
	plan: SubscriptionPlan;
	payment_method_id?: string;
}

export interface UpdateSubscriptionRequest {
	plan?: SubscriptionPlan;
}

export interface CancelSubscriptionRequest {
	reason?: string;
	cancel_at_period_end?: boolean;
}

export interface AddPaymentMethodRequest {
	payment_method_id: string;
	set_as_default?: boolean;
}

export interface SubscriptionSummary {
	subscription: Subscription;
	features: SubscriptionFeatures;
	usage: SubscriptionUsage;
	payment_methods: PaymentMethod[];
	recent_invoices: Invoice[];
}

export interface BillingPortalRequest {
	return_url?: string;
}

export interface BillingPortalResponse {
	url: string;
}

// Plan configuration
export interface PlanConfig {
	name: string;
	price: number;
	currency: string;
	interval: 'month' | 'year';
	features: SubscriptionFeatures;
	popular?: boolean;
	description: string;
	stripe_price_id?: string;
}

export const SUBSCRIPTION_PLANS: Record<SubscriptionPlan, PlanConfig> = {
	free: {
		name: 'Free',
		price: 0,
		currency: 'USD',
		interval: 'month',
		description: 'Perfect for getting started with personal goal tracking',
		features: {
			personal_goals_limit: 3,
			can_join_groups: false,
			group_limit: 0,
			advanced_analytics: false,
			export_data: false,
			priority_support: false,
			custom_categories: false,
			api_access: false
		}
	},
	premium: {
		name: 'Premium',
		price: 9.99,
		currency: 'USD',
		interval: 'month',
		description: 'Unlock the full power of group accountability and advanced features',
		popular: true,
		features: {
			personal_goals_limit: -1, // Unlimited
			can_join_groups: true,
			group_limit: 10,
			advanced_analytics: true,
			export_data: true,
			priority_support: true,
			custom_categories: true,
			api_access: true
		},
		stripe_price_id: 'price_premium_monthly' // Will be set from environment
	}
};

// Helper functions
export const getPlanConfig = (plan: SubscriptionPlan): PlanConfig => {
	return SUBSCRIPTION_PLANS[plan];
};

export const getSubscriptionFeatures = (plan: SubscriptionPlan): SubscriptionFeatures => {
	return SUBSCRIPTION_PLANS[plan].features;
};

export const isActive = (subscription: Subscription): boolean => {
	return subscription.status === 'active' || subscription.status === 'trial';
};

export const isInTrial = (subscription: Subscription): boolean => {
	if (!subscription.trial_end) return false;
	
	const trialEnd = new Date(subscription.trial_end);
	const now = new Date();
	
	return subscription.status === 'trial' && now < trialEnd;
};

export const isPremium = (subscription: Subscription): boolean => {
	return subscription.plan === 'premium' && isActive(subscription);
};

export const getDaysRemainingInTrial = (subscription: Subscription): number => {
	if (!isInTrial(subscription) || !subscription.trial_end) return 0;
	
	const trialEnd = new Date(subscription.trial_end);
	const now = new Date();
	const diffTime = trialEnd.getTime() - now.getTime();
	const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
	
	return Math.max(diffDays, 0);
};

export const canCreatePersonalGoal = (subscription: Subscription, currentGoalsCount: number): boolean => {
	const features = getSubscriptionFeatures(subscription.plan);
	
	if (features.personal_goals_limit === -1) return true; // Unlimited
	
	return currentGoalsCount < features.personal_goals_limit;
};

export const canJoinGroups = (subscription: Subscription): boolean => {
	const features = getSubscriptionFeatures(subscription.plan);
	return features.can_join_groups && isActive(subscription);
};

export const hasAdvancedAnalytics = (subscription: Subscription): boolean => {
	const features = getSubscriptionFeatures(subscription.plan);
	return features.advanced_analytics && isActive(subscription);
};

export const getSubscriptionStatusColor = (status: SubscriptionStatus): string => {
	switch (status) {
		case 'trial':
			return 'text-blue-600 bg-blue-100';
		case 'active':
			return 'text-green-600 bg-green-100';
		case 'past_due':
			return 'text-yellow-600 bg-yellow-100';
		case 'canceled':
			return 'text-gray-600 bg-gray-100';
		case 'unpaid':
			return 'text-red-600 bg-red-100';
		default:
			return 'text-gray-600 bg-gray-100';
	}
};

export const getSubscriptionStatusMessage = (subscription: Subscription): string => {
	switch (subscription.status) {
		case 'trial':
			const daysLeft = getDaysRemainingInTrial(subscription);
			return `${daysLeft} day${daysLeft !== 1 ? 's' : ''} left in trial`;
		case 'active':
			return 'Active subscription';
		case 'past_due':
			return 'Payment past due';
		case 'canceled':
			return 'Subscription canceled';
		case 'unpaid':
			return 'Payment required';
		default:
			return 'Unknown status';
	}
};

export const getPlanPrice = (plan: SubscriptionPlan): string => {
	const config = getPlanConfig(plan);
	
	if (config.price === 0) return 'Free';
	
	return `$${config.price}/${config.interval}`;
};

export const formatPrice = (amount: number, currency: string = 'USD'): string => {
	return new Intl.NumberFormat('en-US', {
		style: 'currency',
		currency: currency
	}).format(amount);
};

export const getUsagePercentage = (current: number, limit: number): number => {
	if (limit === -1) return 0; // Unlimited
	if (limit === 0) return 100;
	
	return Math.min((current / limit) * 100, 100);
};

export const getUsageColor = (percentage: number): string => {
	if (percentage >= 100) return 'text-red-600 bg-red-100';
	if (percentage >= 80) return 'text-orange-600 bg-orange-100';
	if (percentage >= 60) return 'text-yellow-600 bg-yellow-100';
	return 'text-green-600 bg-green-100';
};

export const shouldShowUpgradePrompt = (subscription: Subscription, feature: keyof SubscriptionFeatures): boolean => {
	if (isPremium(subscription)) return false;
	
	const features = getSubscriptionFeatures(subscription.plan);
	return !features[feature];
};

export const getUpgradeMessage = (feature: string): string => {
	const messages: Record<string, string> = {
		'groups': 'Join accountability groups with Premium! 🤝',
		'analytics': 'Get detailed analytics with Premium! 📊',
		'unlimited_goals': 'Create unlimited goals with Premium! 🎯',
		'export': 'Export your data with Premium! 📥',
		'api': 'Access our API with Premium! 🔌',
		'support': 'Get priority support with Premium! 💬'
	};
	
	return messages[feature] || 'Upgrade to Premium for more features! ✨';
};

export const getCardBrandColor = (brand: string): string => {
	switch (brand.toLowerCase()) {
		case 'visa':
			return 'text-blue-600';
		case 'mastercard':
			return 'text-red-600';
		case 'amex':
		case 'american_express':
			return 'text-green-600';
		case 'discover':
			return 'text-orange-600';
		default:
			return 'text-gray-600';
	}
};

export const getCardBrandIcon = (brand: string): string => {
	switch (brand.toLowerCase()) {
		case 'visa':
			return '💳';
		case 'mastercard':
			return '💳';
		case 'amex':
		case 'american_express':
			return '💳';
		case 'discover':
			return '💳';
		default:
			return '💳';
	}
};